### DISCUSSION TITLE
Revive 2.1 Dev Log

### DISCUSSION THREAD

**Discussion by MXKhronos** [HelixNebulaStudio/RiseOfTheDead#355] [Announcements]
# Revive 2.1 Dev Log

I'm going to start writing dev logs for each updates and hopefully stick to it for future updates. These will give you insights on parts of the updates I'm open to discuss about.

# Development
I'm changing up some things in my workflow, a large, feature packed, content packed bundle of an update is too taxing for me and probably too much to juggle. I'm going to go back to another way of releasing updates, where feature, content are smaller and released as soon as they are done testing. Like this commit, https://github.com/HelixNebulaStudio/RiseOfTheDead/commit/e566cc7d68dbd15ba61c0cd9b3cdd447c5021444 will be put into update 2.2's change logs and is for 2.2 but technically already live.

# Fumes Rework
Old Fumes model is hideous, I made it when I barely know how to 3D model. His abilities aren't complex and I'd like to keep it that way.

He vents out a cloud of fumes, blocking vision in and out. Inside the `Fumes Cloud` is also pretty dense. 
- The idea is that `Fumes` isn't aggressive or defensive, he's just simply there. He is a strong denial of area. Invisible to threat-sense and weak points due to the fumes, you will either have to go into his cloud or predict where he is from the outside to eliminate him.
- Added mechanics for `Fumes Cloud` to be able to damage your `Gas Protection` gear is the cost of entering his cloud.
- With the right gear, you will be able to stay inside his cloud for longer. In normal mode, he only stands in a specific spot, you have a good amount of time to find and kill him.
- In hard mode, `Fumes` will wander around inside his cloud. Predicting his location from the outside would be much harder. You will have to enter the cloud in order to find him. Your `Gas Protection` gear will also take more damage, reducing the time you have to find him.
- Or you can prepare a bunch of `gasmask` to burn through, if you don't mind the cost of repairing them.

# Npc Tasks, Inventory and Stats
This has been something I've been wanting to work on for a long time. The only reason I actually got started was because I wanted to a passive way for players to obtain `colorcustom` of a color they want.
- It couldn't be a board mission because once you've obtained all the colors, there's little reason do it again.
- I also did not want to add a mission where players need to scavenge for it themselves, that concept's done before.
- So it had to be done by an Npc, hence the mission `The Killer Hues` needed a Npc task and inventory system. 

## Npc Inventory
Base on the weapon modding system where there can only be 5 mods, the Npc inventory is similar where there can only be 5 items.
However, the item types is a bit more restricting where there's now a primary, secondary, melee, clothing and a miscellaneous slot to keep things balanced and feasible to implement.

Currently, they don't really do anything with their inventory items. If it's a weapon, they will equip it for aesthetics. In `The Killer Hues`, `Lydia` would use the weapon that's equipped, that was made to test out the system.

In a way, they are basically extra storage space for now, you can not kick an Npc with items however. 
## Npc Tasks
Npc tasks opens up some possibilities and gives safehome Npcs some use. To keep things familiar, it operates like the workbench. Currently, all Npcs can only do one thing at a time, but future plans to introduce ways to level them up to do more tasks at the same time. But having a maximum of 5 safehome Npcs which could do 5 tasks at a time is already plenty for now.

Completing `The Killer Hues` unlocks the task `Scavenge Colors`. It is the first task added, in which you can pick a color you want `Lydia` to scavenge for.

Another task added for all Npcs is, `Scavenge Food`. It's a basic task to scavenge for food, food is picked from the drop table, `npctask:foodscavenge`. Adds another way to obtain some rare foods like `sandwich` and `energydrink`.

Am definitely open for ideas for what Npc tasks you would like to see.. And then I'll consider what's feasible to add.
## Npc Stats
Since Npcs didn't have much purpose prior to the update, their `Hunger` and `Happiness` stat didn't matter and was probably bugged or something..

Anyways, it is now reworked, `Hunger` now drains slowly, about 28.8% per day, but a single quantity of a food item stored in the safehome freezer would restore 30%.

- `Health` currently have no effect and always remains at 100%. 
- `Armor` currently have no effect and always remains at 10 ap.

`Happiness` is the weighted sum of `Hunger` and `Health`.

The purpose of these stats are to give you a task of maintaining food supply for your camp of survivors. By keeping them happy, you can assign task that are hard or unpleasant to do. By keeping them fed, they will have a higher chance of not failing.
# The Killer Hues
Side Mission: `Lydia`, a safehome Npc who's just an ordinary psychic survivor trying to survive needs you to teach her out to fend off zombies/hostiles.

`Lydia` has a unique safehome npc class: `FortuneTeller`, which initially was a cosmetic class where she response to your questions with 8ball quotes so I thought it would be fitting to give her a task that is about cosmetics. Scavenging a custom color.
# 5 Years Anniversary Event Pass
Pretty much a event pass for celebrating the 5 years anniversary of Rise of the Dead Revived since the rework of Rise of the Dead Legacy which broke after the filtering update. Play the game and you will earn books to level up the event pass for rewards.
I'm pretty proud of the Fancy Skins Pack I made for it, having seen community weapon skins using it, they managed to make really sick looking designs.
# Event Puzzle
This is for a side project where you can code within the game using ZSharp. As I developer, I do get bored of working on one thing for long periods, working on ZSharp is sorta like taking a break from working on Rise of the Dead to work on something new. But I also want to see what people can do with what I make, this combines both worlds for the enthusiasts out there.

If you had read ZSharpScript's github page, it talks about the potential of modding or server "plugins" that this could provide.. That would likely not happening in Rise of the Dead because it's not designed as a sandbox game. As for my other project "Almonds", which is planned to be a sandbox survival that private servers owned by players could have custom ZSharp scripts to modify the game's vanilla experience.

So the puzzle is made for players to dabble and play around with ZSharp, and if you can solve the puzzle, you will be rewarded with 10 event pass levels.
# Unlockables to Skin Perm
I'll admit it, I didn't put too much thought into the change and importance of what players paid to unlock and I was made aware a while after the update. I was more focused on the point where unlockables that weren't meant to be easily obtained such as `zriceraskullinferno` and `vexglovesinferno` and I didn't want the new `gasmasksovietfumes` to simply be cloned and given away. But that's a flaw of unlockables and giving out 2 skin charges for each unlocked unlockable was not suffice. So all unlocked unlockables prior to the update will remain as is with a infinity symbol shown on the workbench, you will be able to apply them unlimited amount of times. But for all skins in the future, they are one time use to apply on a single clothing or weapon.

The change is also to simplify maintaining the system of skins in the backend as more new skins are added, some parts of the system are rewritten for scaling and to remove old *rusty* mechanics. Why not make exceptions? Well, that adds more to maintain and that's why I chose the easier option and made unlockables deprecated. So all new skins will use the new system and .

# Color Picker
In the Npc task `Scavenge Colors`, a player can pick a color for `Lydia` to scavenge. To simplify the process whilst also provide a coherent selection screen for a reasonable amount of colors, I had to redesign the old color picker. This new color picker will be used in more color selection options such as the workbench appearance menu redesign that I really want to work on.

If part of the text is incoherent, that's because I probably only have 3-5 brain cells active atm.
Khronos~

**Comment by GregTheBestest** [comment]
Really good dev log. Thank you, Khronos.

**Comment by Leroy8888** [comment]
Alright, thank you for the explanation and insight on future plans.
